src/extras/reload.js

```diff
@@ -0,0 +1,69 @@
+(function (global) {
+  global.TRACING = true;
+
+  const systemJSPrototype = System.constructor.prototype;
+  const dependentsForModule = {};
+
+  systemJSPrototype.reload = function (id, parentUrl) {
+    const loader = this;
+
+    // An import first sets up a promise and resolves the module id, recreating
+    // the same behavior here avoids a race condition if a reload is called within
+    // that period
+    return Promise.resolve()
+      .then(function () {
+        return loader.resolve(id, parentUrl);
+      })
+      .then(function (id) {
+        const oldLoad = loader.getLoad(id);
+        if (!oldLoad) {
+          // module was not loaded before
+          return loader.import(id);
+        }
+
+        // module was imported before, and it might still be in a loading state
+        // first do a regular import for this module to ensure a current executing load
+        // is finished. except on error, as that will throw the same error again
+        const init = !oldLoad.er ? loader.import(id) : Promise.resolve();
+        return init
+          .catch(function () {
+            // ignore errors so that the reload can recover an error state
+          })
+          .then(function () {
+            // delete module from registry, and re-fetch it while maintaining importerSetters
+            loader.delete(id);
```

Comment:
What if we supported `delete` returning a function to reinject the module?

Something like - `const update = loader.delete(id)` and then later on after the import, `update(moduleInstance)`?

Reply:
To reload a module, we need to:

**Delete the old module** 
**Import the new module**
**Register and run the `importerSetters`**
As far as I can tell the only way to hook into this right now is inside the closure of `getOrCreateLoad` because that's where the export function is created and called for the initial values. Would you move this to the update function?
**Update references to the module in the load wrapper of dependents**
Would you move this to the update function as well?

Reply:
I was imagining the update function running after the module has been `System.import`-ed again, so that its function is just to patch up the previous stale parent references.

Let's have the update function take no argument, and just assume a module of the same id is now present in the registry.

importerSetters is the list of setter functions of the parents.

The update function would then run the old importerSetters against the new module, and then it also needs to update the importerSetters of the reloaded module to include the importer setters of the parents that weren't reloaded (as these won't be in the module's importerSetters list). It's some careful stitching but that is the game here.

Reply:
That would work, though we do need to pass in the dependents. 

This moves most of the code from the reload extra into the registry, from a code structure perspective it seems like it's easier easier to add a reload function to the registry. Not sure where you draw the line in terms of filesize.

Reply:
The previous importerSetters list should be enough to do the patching I think.

Note that `delete` should itself remove the module from the `importerSetters` list so that this process doesn't leak memory references (eg a module that imports itself would have already removed itself from the old importerSetters list held by the update function closure).